Namelist Refactor

jjuyeonkim opened the pull request:
**Description**
This is one of multiple PRs to refactor the NDSL's `Namelist`and `NamelistDefaults` so that submodules have more ownership over configuration parameters and corresponding default values.   [NDSL Issue#64](https://github.com/NOAA-GFDL/NDSL/issues/64)

Changes include the following:
- Removes use of `NamelistDefaults` in the physics config. Instead, the config class will store the default values directly.
- Uses simplified `ndsl.Namelist`
- Replaces `namelist` with `config` in `TranslatePhysics`
- Updates translate tests with these changes
- Modifes github workflow tests to run jjuyeonkim's fork's tests until namelist refactor is completed
 
Fixes: https://github.com/NOAA-GFDL/NDSL/issues/64 (refactor that affects Pace, [NDSL](https://github.com/NOAA-GFDL/NDSL/pull/218), [PyFV3](https://github.com/NOAA-GFDL/PyFV3/pull/83), and PySHiELD).

**How Has This Been Tested?**
PyFV3 and PySHiELD translate tests run as expected compared to the `develop` branch. Pace unit tests also pass as expected.

**Checklist:**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published in downstream modules  -- **_There are changes in Pace, [NDSL](https://github.com/NOAA-GFDL/NDSL/pull/218), [PyFV3](https://github.com/NOAA-GFDL/PyFV3/pull/83), and PySHiELD to address this issue. They need to be checked in together._**
- [ ] New check tests, if applicable, are included
- [ ] Targeted model, if this change was triggered by a model need/shortcoming

jjuyeonkim committed:
Namelist Refactor: NDSL Issue#64
- Removing NamelistDefaults in the physics config
- Updating to use simplified ndsl.Namelist
- Replacing namelist with config for TranslatePhysicsFortranData2Py
- Modifying workflow tests until namelist refactor is completed

jjuyeonkim committed:
PhysicsConfig fixes; pre-commit checks + linting

jjuyeonkim committed:
Still linting

jjuyeonkim committed:
[TEST] Pointing pyproject.toml's ndsl, pyfv3 to jjuyeonkim fork for CI

jjuyeonkim committed:
Merge branch 'develop' into 20250908_namelist_2

fmalatino commented on pyshield/_config.py:
```diff
@@ -28,84 +33,80 @@ class PhysicsConfig:
+    const_vi: bool = False  # Fall velocity tuning constant of ice
+    const_vr: bool = False  # Fall velocity tuning constant of rain water
+    const_vs: bool = False  # Fall velocity tuning constant of snow
+    vs_fac: float = 1.0  # if const_vs: 1.
```
Same question here about the floats; should they be type hinted as NDSL `Float` or Python native?

oelbert replied:
Not in this PR, because typehinting them as NDSL types breaks things. We'll have to go through at some point and fix the precision on all of these config settings but that'll be a separate set of PRs throughout Pace, PySHiELD, and PyFV3

oelbert commented on pyshield/_config.py:
```diff
@@ -28,84 +33,80 @@ class PhysicsConfig:
     nwat: int = DEFAULT_INT
     schemes: List = None
     do_qa: bool = DEFAULT_BOOL
-    c_cracw: float = NamelistDefaults.c_cracw
```
It may be good to have these live in a local pyshield namelist defaults class but that's a discussion to have with NASA about code layout

oelbert commented on pyshield/_config.py:
```diff
@@ -127,76 +128,33 @@ def __post_init__(self):
+            for key, value in namelist_dict.items()
+            if key in cls.__dataclass_fields__  # type: ignore
+        }
+        return cls(**namelist_dict)
```
I think properly these should be reversed; `from_namelist` should do the lines from 150-160, and `from_f90nml` should set `namelist=Namelist(f90_namelist)` and then call `from_namelist`

jjuyeonkim replied:
This makes sense. I switched them.

oelbert commented on pyshield/_config.py:
```diff
@@ -4,13 +4,18 @@
+    "coupler_nml",
+    "gfdl_cloud_microphysics_nml",
+    "integ_phys_nml",
+)  # TODO Double-check for any additional groups
```
This should include `gfs_physics_nml` to capture more of the parameters

jjuyeonkim replied:
Thanks! I added it.

oelbert reviewed (changes requested):
Looks good, a couple of stylistic changes for now and one group that should be a default.

jjuyeonkim committed:
[TEST] jjuyeonkim ref CI tweaks for NDSL Issue#64

jjuyeonkim committed:
typo fix

jjuyeonkim committed:
Temporary repo/ref for NDSL Issue#64
Simplfying checkouts in translate tests for temporary jjuyeonkim fork workflows

jjuyeonkim committed:
Merge branch 'develop' into 20250908_namelist_2

jjuyeonkim committed:
NDSL Issue#64: Adding gfs_physics_nml to default namelist groups; switching from_namelist/from_f90nml implementation
